src/tools/rbd_mirror/image_sync/ObjectCopyRequest.cc

```diff
@@ -175,22 +175,46 @@ void ObjectCopyRequest<I>::send_write_object() {
   auto &sync_ops = m_snap_sync_ops.begin()->second;
   assert(!sync_ops.empty());
   uint64_t buffer_offset;
+  uint64_t sparse_read_offset;
   librados::ObjectWriteOperation op;
   for (auto &sync_op : sync_ops) {
     switch (std::get<0>(sync_op)) {
     case SYNC_OP_TYPE_WRITE:
       buffer_offset = 0;
-      for(auto it : std::get<4>(sync_op)){
-         bufferlist tmpbl;
-         tmpbl.substr_of(std::get<3>(sync_op), buffer_offset, it.second);
-         op.write(it.first, tmpbl);
-         op.set_op_flags2(LIBRADOS_OP_FLAG_FADVISE_SEQUENTIAL |
-                          LIBRADOS_OP_FLAG_FADVISE_NOCACHE);
-         buffer_offset += it.second;
-	 dout(20) << ": write op: " << it.first<< "~" << it.second << dendl;
+      sparse_read_offset = std::get<1>(sync_op);
+      for (auto it : std::get<4>(sync_op)) {
```

Comment:
I was thinking you could just use the interval_set, initialize one to the requested offset + length, subtract out the extents within the sparse read result, and issue a zero / truncate ops for the remaining extents. The truncate would only (possibly) apply to the last extent if its offset + length == the object size.